### ISSUE TITLE
out of bounds index for exotic parameters

### ISSUE THREAD

**Issue by sandreza** [glwagner/PriceWellerPinkel.jl#18]
When varying parameters in KPP the following error arises

attempt to access OffsetArray(::Array{Float64,1}, 0:129) with eltype Float64 with indices 0:129 at index [130]

To replicate this multiply all the "model constants" in the KPP example by 10. This could mean that the model is not correct in that parameter regime, but to automate parameter optimization it would be better to return NaNs. A potential fix may be to turn off bounds checking.

**Comment by glwagner** [comment]
Awesome, thanks for finding a bug!

I think that if we’re getting this kind of error, turning off bounds checking will just lead to a segfault. Arrays cannot be accessed out of bounds. 

The problem is probably something in the function that determines the mixing depth. 

Can you paste the whole error? You can use triple backticks to format something as code to make it easier to read. 

Also, pasting the script that produced the error will accelerate fixing the bug.

**Comment by glwagner** [comment]
Thinking further this should never happen, so it’s definitely a bug. Index 128 is the “surface” for a model with n=128 in OceanTurb; index 129 is the ghost cell above the surface, and 130 does not exist. This error likely means that some function attempted to calculate a derivative across the face above the ghost cell — which is nonsensical since that’s not in the model domain.

**Comment by sandreza** [comment]
The entire error is below

```
BoundsError: attempt to access OffsetArray(::Array{Float64,1}, 0:129) with eltype Float64 with indices 0:129 at index [130]

Stacktrace:
 [1] throw_boundserror(::OffsetArrays.OffsetArray{Float64,1,Array{Float64,1}}, ::Tuple{Int64}) at ./abstractarray.jl:484
 [2] checkbounds at ./abstractarray.jl:449 [inlined]
 [3] getindex at /Users/andresouza/.julia/packages/OffsetArrays/ae3ez/src/OffsetArrays.jl:135 [inlined]
 [4] getindex at /Users/andresouza/Desktop/OceanTurb.jl-master/src/fields.jl:139 [inlined]
 [5] surface_layer_average(::Field{Cell,OffsetArrays.OffsetArray{Float64,1,Array{Float64,1}},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}, ::Float64, ::Int64) at /Users/andresouza/Desktop/OceanTurb.jl-master/src/models/k_profile_parameterization.jl:169
 [6] Δ at /Users/andresouza/Desktop/OceanTurb.jl-master/src/models/k_profile_parameterization.jl:187 [inlined]
 [7] bulk_richardson_number(::Field{Cell,OffsetArrays.OffsetArray{Float64,1,Array{Float64,1}},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}, ::Field{Cell,OffsetArrays.OffsetArray{Float64,1,Array{Float64,1}},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}, ::Field{Cell,OffsetArrays.OffsetArray{Float64,1,Array{Float64,1}},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}, ::Field{Cell,OffsetArrays.OffsetArray{Float64,1,Array{Float64,1}},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}, ::Float64, ::Float64, ::Float64, ::Float64, ::Float64, ::Float64, ::Float64, ::Int64) at /Users/andresouza/Desktop/OceanTurb.jl-master/src/models/k_profile_parameterization.jl:202
 [8] bulk_richardson_number(::OceanTurb.KPP.Model{ForwardEulerTimestepper{OceanTurb.KPP.Solution},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}},Float64}, ::Int64) at /Users/andresouza/Desktop/OceanTurb.jl-master/src/models/k_profile_parameterization.jl:215
 [9] mixing_depth(::OceanTurb.KPP.Model{ForwardEulerTimestepper{OceanTurb.KPP.Solution},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}},Float64}) at /Users/andresouza/Desktop/OceanTurb.jl-master/src/models/k_profile_parameterization.jl:231
 [10] update_state!(::OceanTurb.KPP.Model{ForwardEulerTimestepper{OceanTurb.KPP.Solution},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}},Float64}) at /Users/andresouza/Desktop/OceanTurb.jl-master/src/models/k_profile_parameterization.jl:102
 [11] calc_rhs_explicit!(::OceanTurb.KPP.Solution, ::OceanTurb.KPP.Model{ForwardEulerTimestepper{OceanTurb.KPP.Solution},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}},Float64}) at /Users/andresouza/Desktop/OceanTurb.jl-master/src/models/k_profile_parameterization.jl:357
 [12] iterate!(::OceanTurb.KPP.Model{ForwardEulerTimestepper{OceanTurb.KPP.Solution},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}},Float64}, ::Int64) at /Users/andresouza/Desktop/OceanTurb.jl-master/src/timesteppers.jl:64
 [13] iterate!(::OceanTurb.KPP.Model{ForwardEulerTimestepper{OceanTurb.KPP.Solution},UniformGrid{Float64,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}},Float64}, ::Int64, ::Int64) at /Users/andresouza/Desktop/OceanTurb.jl-master/src/timesteppers.jl:26
 [14] loss_function(::Array{Float64,1}, ::Array{Any,1}) at ./In[57]:43
 [15] top-level scope at ./In[161]:32
```

**Comment by glwagner** [comment]
I think I understand the bug now. This could occur if the parameter `CRi` is set to something below zero (we should always have `CRi > 0`). The problem is that I assumed the condition `Ri₁ < CRi` condition is always true at the beginning of the loop:

https://github.com/glwagner/OceanTurb.jl/blob/f6459a9a47f6b75367ed6acbebf971fe704b1117/src/models/k_profile_parameterization.jl#L229

I didn't allow for the possibility that `CRi < 0`, which would mean that the surface value of `Ri₁=0` is not subcritical... !

I can fix the bug by relaxing this assumption: if for some reason the critical Richardson number criterion is satisfied at the *very top* of the domain, I'll assume the mixing depth is 0. It can't be any shallower than 0. I'll implement this in the next PR.

**Comment by glwagner** [comment]
I believe this is resolved by #21, since 

https://github.com/glwagner/OceanTurb.jl/blob/f598dece5c47a51e5fb50cb091ae0e68de162ed3/src/models/k_profile_parameterization.jl#L242

in `mixing_depth` should ensure that the mixing depth lies within the domain.
